src/debugger/mock-ui/session.js

```diff
@@ -0,0 +1,239 @@
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+import readline from "readline";
+import child_process from "child_process";
+import * as DebugProtocol from "vscode-debugprotocol";
+
+//separator for messages according to the protocol
+const TWO_CRLF = "\r\n\r\n";
+
+export class Session {
+  constructor(proc: Process, adapterPath: string, prepackCommand: string) {
+    this._proc = proc;
+    this._adapterPath = adapterPath;
+    this._prepackCommand = prepackCommand;
+    this._rawData = new Buffer(0);
+    this._contentLength = -1;
+    this._sequenceNum = 0;
+    this._invalidCount = 0;
+  }
+  // the parent (i.e. ui) process
+  _proc: Process;
+  //path to the debug adapter
+  _adapterPath: string;
+  // the child (i.e. adapter) process
+  _adapterProcess: child_process.ChildProcess;
+
+  // buffer to hold incoming data from the adapter
+  _rawData: Buffer;
+  // expected content length for a message
+  _contentLength: number;
+  // id number for each message sent
+  _sequenceNum: number;
+  // interface to read in input from the CLI client
+  _reader: readline.Interface;
+  // number of invalid commands
+  _invalidCount: number;
+  // command to start Prepack with
+  _prepackCommand: string;
+
+  _startAdapter() {
+    let adapterArgs = [this._adapterPath, "--prepack", this._prepackCommand];
+    this._adapterProcess = child_process.spawn("node", adapterArgs);
+    this._proc.on("exit", () => {
+      this.shutdown();
+    });
+    this._proc.on("SIGINT", () => {
+      this.shutdown();
+    });
+    this._adapterProcess.on("exit", () => {
+      this.shutdown();
+    });
+    this._adapterProcess.stdout.on("data", (data: Buffer) => {
+      this._handleData(data);
+    });
+    this._adapterProcess.stderr.on("data", (data: Buffer) => {
+      console.error(data.toString());
+      this.shutdown();
+    });
+  }
+
+  // callback to handle data written back by the adapter
+  _handleData(data: Buffer): void {
+    this._rawData = Buffer.concat([this._rawData, data]);
+    // the following code parses a message according to the protocol.
+    while (this._rawData.length > 0) {
+      // if we know what length we are expecting
+      if (this._contentLength >= 0) {
+        // we have enough data to check for the expected message
+        if (this._rawData.length >= this._contentLength) {
+          // first get the expected message
+          let message = this._rawData.toString("utf8", 0, this._contentLength);
+          // reduce the buffer by the message we got
+          this._rawData = this._rawData.slice(this._contentLength);
+          // reset the content length to ensure it is extracted for the next message
+          this._contentLength = -1;
+          // process the message
+          if (message.length > 0) {
```

Comment:
a comment to explain why the message could be length 0 would be helpful.

Reply:
It should not be, I removed this check.